### DISCUSSION TITLE
Long Button Press Detection

### DISCUSSION THREAD

**Discussion by GeorgeIoak** [Bodmer/TFT_eSPI#1730] [Q&A - Touch]
Has anyone done something with detecting long button presses. I'm trying to create a UI for a thermostat that has UP/DOWN buttons and a tap will change by 1 but a hold of more than 1 second will change by 10.

I was experimenting with GUIslice and had something working but I'm thinking of abandoning GUIslice and just using TFT_eSPI for my simple UI.

Here's my test code for right now (based on the https://github.com/Bodmer/TFT_eSPI/blob/master/examples/Generic/TFT_Button_Label_Datum/TFT_Button_Label_Datum.ino example)
```
// Adjust press state of each key appropriately
  for (uint8_t b = 0; b < NUM_KEYS; b++)
  {
    if (pressed && key[b].contains(t_x, t_y))
    {
      key[b].press(true); // tell the button it is pressed
      if (millis() - timePressStart > 1000)
      {
        switch (b)
        {
          case 0:  // UP Button
            if (desiredTemp <= MAX_H2O_TEMP)
              desiredTemp += TEMP_INCREMENT;
            else desiredTemp = MAX_H2O_TEMP;
          case 1:  // DOWN Button
            if (desiredTemp >= MIN_H2O_TEMP)
              desiredTemp += TEMP_INCREMENT;
            else
              desiredTemp = MIN_H2O_TEMP;
          case 2:  // RESET Button
            desiredTemp = DEFAULT_TEMP;
          default:
            break;
        }
        timePressStart = millis();  // Reset button press time
      }
      Serial.println("Button held more than 1 sec");
      tft.setTextColor(TFT_BLUE, TFT_BLACK);
      tft.drawNumber(desiredTemp, 240, 230, 8);
    }
    else
      key[b].press(false); // tell the button it is NOT pressed
  }

  // Check if any key has changed state
  for (uint8_t b = 0; b < NUM_KEYS; b++)
  {
    // If button was just pressed, redraw inverted button
    if (key[b].justPressed())
    {
      timePressStart = millis();
      pressed = true;
      switch (b)
      {
        case 0:  // UP Button
          desiredTemp++;
          break;
        case 1:  // DOWN Button
          desiredTemp--;
          break;
        case 2:
          break;  // Need to hold for 1 second to reset
      }
      Serial.println("Button " + (String)b + " pressed");
      key[b].drawButton(true, "BUTTON " + (String)(b));
    }

    // If button was just released, redraw normal color button
    if (key[b].justReleased())
    {
      Serial.println("Button " + (String)b + " released");
      key[b].drawButton(false, "BUTTON " + (String)(b));
      pressed = false;
    }
  }
```

But I'm getting my debug statement of `"Button held more than 1 sec"` coming through repeatedly rather than only after 1 second. I'm thinking maybe `key[b].press(true)` is resetting the timer when it changes the button state but I'm not sure.

Before I start digging in I thought I'd ask since I'm sure others have handled long button presses before?

**Comment by Bodmer** [comment]
I think the problem is that justPressed() etc looks at the history of the past 2 press states [stored in the library](https://github.com/Bodmer/TFT_eSPI/blob/master/Extensions/Button.cpp#L97-L105) to look for a change.

This means you must keep that history up to date in the first for loop by ussing key[b].pressed(true) if it has been pressed and key[b].pressed(false) if it is not.

**Reply by Bodmer** [reply]
Ah, I see that you have done that. You also need to update the history if justPressed is true by adding x.pressed(true),  so the library history record is updated. You might think that would be automatic, so this is one thing I dislike about the implementation simplicity of the Adafruit button library, sadly this adds implementeation complexity in the sketch.

**Reply by GeorgeIoak** [reply]
I've rewritten that code a bunch since I posted this and didn't get it to work properly but since you've started work on a new button library I might just give up on this and use your new library since with only a minor modification I do have it reacting if you have the button HELD.

But, before I walk away from this, when you said `update the history if justPressed is true by adding x.pressed(true)` you really mean `x.pressed(true) and not `x.press(true)` because I only have `x.press(true)` in the code above (and for those reading along x is short for key[b])

**Reply by Bodmer** [reply]
Yep, typo, press() not pressed() which is not a library function!

If you look at the library [code here](https://github.com/Bodmer/TFT_eSPI/blob/master/Extensions/Button.cpp#L97-L105) you can see how it determines the button state changes  from the laststate and currstate variables that are set/reset using the press() function.
